snapshot/manifest.go

```diff
@@ -91,6 +92,12 @@ type HasDirEntry interface {
 	DirEntry() *DirEntry
 }
 
+// HasDirEntryFromPlaceholder is implemented by objects that may have a DirEntry
```

Comment:
can we call this `HasDirEntry` ? In principle `repofs` can also implement `HasDirEntry`, right and there are no placeholders involved.

Reply:
I spent a while trying to merge `HasDirEntryFromPlaceholder` and `HasDirEntry` and decided that maintaining the separation is nicer. Summary: 

* `HasDirEntry` is satisfied by *any* `repofs` instances that have a `DirEntry`
* `HasDirEntryFromPaceholder` is satisfied by `localfs` instances and when it returns a non-nil `*DirEntry` to the the `Uploader`, it is expected that the `*DirEntry` is from the same repository as the one being uploaded to.

I can unify them but I can't see a particularly nice way to do it. Details:

* Currently only `localfs` and `ignorefs` implement `HasDirEntryFromPlaceholder`
* `ignorefs` just delegates to its contained `fs.Directory`
* (Correct) `localfs` implementers of `HasDirEntryFromPlaceholder` only return a non-nil `*DirEntry` if the placeholder exists and contains a serialized `DirEntry`
* (per your earlier suggestion) code in `upload.go` verifies that this `DirEntry` came from the repository that it's uploading to
* during snapshot migrations, the `ignorefs` instances contain `repofs` instances that don't implement `HasDirEntryFromPlaceholder` but do implement `HasDirEntry`.
* Moreover, these instances are for a *foreign* repository. So the `DirEntry` validation fails.

So, my thoughts:

* I could push the destination repo validation into `DirEntry` for `localfs` implementations but exposing the necessary state from the `*Uploader` to the `localfs` seemed like a layering violation to me
* I could do something  with reflect in `*ignoreDirectory.DirEntry` to figure out if it contains a `repofs` instance but this seems like a type unsafe work around for not having two interfaces. (I wrote this case to prove to myself that I'm understanding what's happening here.)
* I could skip the `DirEntry` repository verification in `Uploader` during a migration but this is ugly.